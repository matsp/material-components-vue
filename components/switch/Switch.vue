<template>
  <div
    :class="classes"
    class="mdc-switch">
    <div class="mdc-switch__track"/>
    <div class="mdc-switch__thumb-underlay">
      <div class="mdc-switch__thumb">
        <input
          v-model="model"
          :disabled="disabled"
          type="checkbox"
          class="mdc-switch__native-control"
          role="switch" >
      </div>
    </div>
  </div>
</template>

<script>
import { MDCSwitch } from '@material/switch'
import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  model: {
    prop: 'checked',
    event: 'change'
  },
  props: {
    checked: {
      type: Boolean,
      required: false
    },
    disabled: {
      type: Boolean,
      required: false
    }
  },
  data () {
    return {
      mdcSwitch: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-switch--disabled': this.disabled
      }
    },
    model: {
      get () {
        return this.checked
      },
      set (state) {
        this.$emit('change', state)
      }
    }
  },
  mounted () {
    this.mdcSwitch = MDCSwitch.attachTo(this.$el)
    this.mdcSwitch.checked = this.checked
    this.mdcSwitch.disabled = this.disabled
  },
  beforeDestroy () {
    this.mdcSwitch.destroy()
  }
}
</script>
