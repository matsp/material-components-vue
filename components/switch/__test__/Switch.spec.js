import { mount } from '@vue/test-utils'
import Switch from '../Switch.vue'

describe('Switch', () => {
  it('should mount', () => {
    const wrapper = mount(Switch)
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.vm.$data.mdcSwitch).toBeDefined()
  })

  it('should render with no prop', () => {
    const wrapper = mount(Switch)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-switch')
    expect(wrapper.find('input').attributes('disabled')).toBeUndefined()
  })

  it('should render as disabled', () => {
    const wrapper = mount(Switch, {
      propsData: {
        disabled: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-switch--disabled')
    expect(wrapper.find('input').attributes('disabled')).toBeDefined()
  })

  it('should render as checked', () => {
    const wrapper = mount(Switch, {
      propsData: {
        checked: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.vm.$data.mdcSwitch.checked).toBeTruthy()
    expect(wrapper.find('input').element.checked).toBeTruthy()
  })

  it('should render and emit', () => {
    const wrapper = mount(Switch)

    const input = wrapper.find('input')

    input.setChecked()
    expect(wrapper.emitted().change).toBeTruthy()
    expect(wrapper.vm.$data.mdcSwitch.checked).toBeTruthy()
    expect(wrapper.find('input').element.checked).toBeTruthy()

    input.setChecked(false)
    expect(wrapper.emitted().change).toBeTruthy()
    expect(wrapper.vm.$data.mdcSwitch.checked).toBeFalsy()
    expect(wrapper.find('input').element.checked).toBeFalsy()
  })
})
