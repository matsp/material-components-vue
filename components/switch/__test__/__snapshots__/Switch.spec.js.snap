// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render as checked 1`] = `
<div class="mdc-switch mdc-switch--checked">
  <div class="mdc-switch__track"></div>
  <div class="mdc-switch__thumb-underlay">
    <div class="mdc-switch__thumb"><input role="switch" type="checkbox" class="mdc-switch__native-control"></div>
  </div>
</div>
`;

exports[`Switch should render as disabled 1`] = `
<div class="mdc-switch mdc-switch--disabled">
  <div class="mdc-switch__track"></div>
  <div class="mdc-switch__thumb-underlay">
    <div class="mdc-switch__thumb"><input disabled="" role="switch" type="checkbox" class="mdc-switch__native-control"></div>
  </div>
</div>
`;

exports[`Switch should render with no prop 1`] = `
<div class="mdc-switch">
  <div class="mdc-switch__track"></div>
  <div class="mdc-switch__thumb-underlay">
    <div class="mdc-switch__thumb"><input role="switch" type="checkbox" class="mdc-switch__native-control"></div>
  </div>
</div>
`;
